server/neptune/workflows/internal/deploy/revision/queue/pull_rebaser.go

```diff
@@ -0,0 +1,159 @@
+package queue
+
+import (
+	"context"
+	"path/filepath"
+	"strings"
+	"time"
+
+	"github.com/docker/docker/pkg/fileutils"
+	"github.com/pkg/errors"
+	key "github.com/runatlantis/atlantis/server/neptune/context"
+	"github.com/runatlantis/atlantis/server/neptune/workflows/activities"
+	"github.com/runatlantis/atlantis/server/neptune/workflows/activities/github"
+	"github.com/runatlantis/atlantis/server/neptune/workflows/activities/terraform"
+	"github.com/runatlantis/atlantis/server/neptune/workflows/internal/config/logger"
+	"go.temporal.io/sdk/workflow"
+)
+
+// GH API Ratelimit resets every hour, 2 hours should be enough for the GH API call failures to autoresolve
+const RebaseGithubScheduleToCloseTimeout = 2 * time.Hour
+
+type githubRebaseActivities interface {
+	GithubListOpenPRs(ctx context.Context, request activities.ListOpenPRsRequest) (activities.ListOpenPRsResponse, error)
+	GithubListModifiedFiles(ctx context.Context, request activities.ListModifiedFilesRequest) (activities.ListModifiedFilesResponse, error)
+}
+
+type prRevisionSetterActivities interface {
+	SetPRRevision(ctx context.Context, request activities.SetPRRevisionRequest) (activities.SetPRRevisionResponse, error)
+}
+
+type PullRebaser struct {
+	GithubActivities           githubRebaseActivities
+	PrRevisionSetterActivities prRevisionSetterActivities
+}
+
+func (p *PullRebaser) RebaseOpenPRsForRoot(ctx workflow.Context, repo github.Repo, root terraform.Root) error {
+	originalOpts := workflow.GetActivityOptions(ctx)
+
+	// Configure github specific ScheduleToClose timeout for GH API call failures to autoresolve
+	// TODO: Add alarms on GithubListOpenPRs and GithubListModifiedFiles API call failures
+	ghOpts := originalOpts
+	ghOpts.ScheduleToCloseTimeout = RebaseGithubScheduleToCloseTimeout
+	ctx = workflow.WithActivityOptions(ctx, ghOpts)
+
+	// list open PRs
+	var listOpenPRsResp activities.ListOpenPRsResponse
+	err := workflow.ExecuteActivity(ctx, p.GithubActivities.GithubListOpenPRs, activities.ListOpenPRsRequest{
+		Repo: repo,
+	}).Get(ctx, &listOpenPRsResp)
+	if err != nil {
+		return errors.Wrap(err, "listing open PRs")
+	}
+
+	// spawn activities to list modified fles in each open PR in parallel
+	futureByPullNum := map[github.PullRequest]workflow.Future{}
+	for _, pullRequest := range listOpenPRsResp.PullRequests {
+		future := workflow.ExecuteActivity(ctx, p.GithubActivities.GithubListModifiedFiles, activities.ListModifiedFilesRequest{
+			Repo:        repo,
+			PullRequest: pullRequest,
+		})
+		futureByPullNum[pullRequest] = future
+	}
+
+	// resolve the futures and check if an open PR needs to be rebased
+	prsToRebase := []github.PullRequest{}
+	for pullRequest, future := range futureByPullNum {
+		var result activities.ListModifiedFilesResponse
+
+		// list modified files should not fail unless we hit the ratelimit which should autoresolve once our ratelimit revives in an hour max
```

Comment:
this phrasing isn't correct, it might autoresolve, but there's exponential backoff here, so it could result in this whole process taking forever.

Additionally, there's other things that could fail such as transient failures, gh outages etc.